### PR TITLE
Ruschein

### PR DESCRIPTION
**Pull request by ruschein**

### CODE REVIEW

File: cpp/lib/include/ProcessUtil.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <string>
+#include <unordered_set>
+#include <sys/types.h>
+
+
+namespace ProcessUtil {
+
+
+/** \return  The list of PID's that have an open file descripto for "path".
```
Comment (mtrojan-ub) [inline cpp/lib/include/ProcessUtil.h:36]: descripto

File: cpp/lib/include/ProcessUtil.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/** \file    ProcessUtil.h
+ *  \brief   The ProcessUtil interface.
+ *  \author  Dr. Johannes Ruscheinski
+ */
+
+/*
+ *  Copyright 2020 Universitätsbibliothek Tübingen
+ *
+ *  This file is part of the libiViaCore package.
```
Comment (mtrojan-ub) [inline cpp/lib/include/ProcessUtil.h:9]: really?

Comment (mtrojan-ub) [reply cpp/lib/include/ProcessUtil.h:9]: please also check the comment blocks below that also reference to libiViaCore

File: cpp/lib/include/ProcessUtil.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/** \file    ProcessUtil.h
+ *  \brief   The ProcessUtil interface.
+ *  \author  Dr. Johannes Ruscheinski
+ */
+
+/*
+ *  Copyright 2020 Universitätsbibliothek Tübingen
```
Comment (mtrojan-ub) [inline cpp/lib/include/ProcessUtil.h:7]: University Library of Tübingen

File: cpp/lib/src/ProcessUtil.cc [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/** \file    ProcessUtil.c
+ *  \brief   The ProcessUtil implementation.
+ *  \author  Dr. Johannes Ruscheinski
+ */
+
+/*
+ *  Copyright 2020 Universitätsbibliothek Tübingen
```
Comment (mtrojan-ub) [inline cpp/lib/src/ProcessUtil.cc:7]: University Library of Tübingen

File: cpp/lib/src/ProcessUtil.cc [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ *  This file is part of the libiViaCore package.
```
Comment (mtrojan-ub) [inline cpp/lib/src/ProcessUtil.cc:9]: Also the whole libiViaCore stuff

Comment (mtrojan-ub) [reply cpp/lib/src/ProcessUtil.cc:9]: By the way, do we really need a license statement in each file? In the meantime it is possible to set a license for a github repository.

File: cpp/test/GetProcessIdsForPath_test.cc [hunk 6]
```diff
@@ -0,0 +1,26 @@
+// Test harness for ProcessUtil::GetProcessIdsForPath.
+#include "FileUtil.h"
+#include "ProcessUtil.h"
+#include "util.h"
+
+
+int Main(int, char **) {
+    const std::string TEMPFILE_NAME("/tmp/GetProcessIdsForPath_test.temp");
```
Comment (mtrojan-ub) [inline cpp/test/GetProcessIdsForPath_test.cc:8]: use FileUtil::AutoTempFile instead?

**Review by mtrojan-ub — CHANGES_REQUESTED** [review]
see comments

File: cpp/test/GetProcessIdsForPath_test.cc (hunk 6 above)
Comment (ruschein) [reply cpp/test/GetProcessIdsForPath_test.cc:8]: I intentionally did not use it to make what's happening more explicit.

File: cpp/lib/src/ProcessUtil.cc (hunk 5 above)
Comment (ruschein) [reply cpp/lib/src/ProcessUtil.cc:9]: Well, maybe we don't

File: cpp/lib/src/ProcessUtil.cc [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace ProcessUtil {
+
+
+std::unordered_set<pid_t> GetProcessIdsForPath(const std::string &path) {
+    std::unordered_set<pid_t> pids;
+
+    FileUtil::Directory proc("/proc", "\\d+"); // Only PID's
+    for (const auto &pid_dir : proc) {
+        FileUtil::Directory fd_dir("/proc/" + pid_dir.getName() + "/fd");
```
Comment (mtrojan-ub) [inline cpp/lib/src/ProcessUtil.cc:35]: due to pid_dir.getName() documentation, it returns the full name of the entry including the directory path, so prefixing it with "/proc" should not be necessary.

File: cpp/lib/src/ProcessUtil.cc [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        for (const auto &link : fd_dir) {
+            std::string link_target;
+            if (FileUtil::ReadLink("/proc/" + pid_dir.getName() + "/fd/" + link.getName(), &link_target) and link_target == path) {
```
Comment (mtrojan-ub) [inline cpp/lib/src/ProcessUtil.cc:38]: Same here

File: cpp/log_rotate.cc [hunk 9]
```diff
@@ -117,6 +127,9 @@ int Main(int argc, char *argv[]) {
                                                FileUtil::GroupnameFromGID(gid));
                     if (SELinuxUtil::IsEnabled())
                         SELinuxUtil::FileContext::ApplyChanges(filename);
+
+                    for (const auto pid : pids)
+                        ::kill(pid, SIGHUP);
```
Comment (mtrojan-ub) [inline cpp/log_rotate.cc:132]: Is it necessary to send SIGHUP also to ourself?
If not, it might be useful if ProcessUtil::GetProcessIdsForPath would take an additional parameter to include or exclude the process itself from the result.

File: cpp/log_rotate.cc [hunk 10]
```diff
@@ -98,10 +101,17 @@ int Main(int argc, char *argv[]) {
...
-            if (max_line_count > 0)
-                FileUtil::OnlyKeepLastNLines(directory_path + "/" + entry.getName(), max_line_count);
-            else {
-                const std::string filename(directory_path + "/" + entry.getName());
+
+            const std::string filename(directory_path + "/" + entry.getName());
+            if (max_line_count > 0) {
+                const int fd(::open(filename.c_str(), O_RDWR));
+                { // New scope to ensure that the FileLocker instance goes out of scope before we close fd!
+                    FileLocker file_locker(fd, FileLocker::WRITE_ONLY);
```
Comment (mtrojan-ub) [inline cpp/log_rotate.cc:109]: FileUtil::OnlyKeepLastNLines opens it with O_RDONLY to calculate no_of_reading_bytes_to_remove, then it passes it on to ::truncate later which needs write access.

FileLocker only supports READ_ONLY and WRITE_ONLY, maybe we need a READ_WRITE flag?

File: cpp/lib/src/FileLocker.cc [hunk 11]
```diff
@@ -34,22 +35,36 @@
...
     struct flock lock_struct;
     lock_struct.l_type   = static_cast<short>((lock_type == READ_ONLY) ? F_RDLCK : F_WRLCK); /* F_RDLCK, F_WRLCK, F_UNLCK    */
     lock_struct.l_whence = SEEK_SET;                                                         /* SEEK_SET, SEEK_CUR, SEEK_END */
     lock_struct.l_start  = 0;                                                                /* Offset from l_whence         */
     lock_struct.l_len    = 0;                                                                /* length, 0 = to EOF           */
     lock_struct.l_pid    = ::getpid();                                                       /* our PID                      */
 
-    if (::fcntl(lock_fd_, F_SETLKW, &lock_struct) == -1)
-        throw std::runtime_error("in FileLocker::FileLocker: fcntl(2) failed! (" + std::to_string(errno) + ")!");
+    if (::fcntl(lock_fd_, F_SETLKW, &lock_struct) == -1) {
```
Comment (mtrojan-ub) [inline cpp/lib/src/FileLocker.cc:53]: From the manpage:
F_SETLKW: As for F_SETLK, but if a conflicting lock is held on the file, then wait for that lock to be released.

I don't see our timeout variable being used here?

File: cpp/log_rotate.cc (hunk 10 above)
Comment (mtrojan-ub) [reply cpp/log_rotate.cc:109]: Or maybe WRITE_ONLY would be correct and should be renamed to READ_WRITE?

File: cpp/lib/src/FileLocker.cc [hunk 12]
```diff
@@ -34,22 +35,36 @@
...
+        const int last_errno(errno);
+        ::alarm(0); // Cancel our alarm!
+        if (last_errno != EINTR)
+            throw std::runtime_error("fcntl(2) failed! (" + std::to_string(last_errno) + ")!");
+        errno = 0;
+        interrupted_ = true;
```
Comment (mtrojan-ub) [inline cpp/lib/src/FileLocker.cc:59]: So this will be used later to prevent an unlock.

But if we get here this also means that the locking has never been complete, so how do we guarantee a file is locked after this function ends? Shouldn't it be called in a loop while the locking hasn't been successful?

File: cpp/lib/src/util.cc [hunk 13]
```diff
@@ -25,15 +25,37 @@
...
+    ::close(log_fd);
+    const int new_fd(::open(log_path.c_str(), O_WRONLY|O_APPEND));
+    if (new_fd == -1) { // We're screwed!
+        const std::string error_message("in HUPHandler(util.cc): can't open(2) \"" + log_path + "\"!\n");
+        ::write(STDERR_FILENO, error_message.c_str(), error_message.size());
+        ::exit(EXIT_FAILURE);
+    }
+
+    // Make sure we're using the same file descriptor as before!
+    ::dup2(new_fd, log_fd);
```
Comment (mtrojan-ub) [inline cpp/lib/src/util.cc:55]: what if ::dup2 fails?

File: cpp/lib/src/util.cc [hunk 14]
```diff
@@ -64,6 +89,13 @@ Logger::Logger()
 }
 
 
+void Logger::redirectOutput(const int new_fd) {
```
Comment (mtrojan-ub) [inline cpp/lib/src/util.cc:92]: In the current version in ruschein branch, redirectOutput is directly implemented in util.h which is not part of this pull request. Did you forget to commit it?

File: cpp/lib/src/util.cc [hunk 15]
```diff
@@ -25,15 +25,37 @@
...
+#include "FileLocker.h"
+#include "FileUtil.h"
 #include "MiscUtil.h"
+#include "SignalUtil.h"
 #include "StringUtil.h"
 #include "TimeUtil.h"
 
 
+static std::string log_path;
+static int log_fd;
```
Comment (mtrojan-ub) [inline cpp/lib/src/util.cc:42]: 1) What's the difference between log_fd and fd_?
2) It might be clearer to either move these variables to util.h (next to fd_) declaration, or to convert fd_ to a static variable as well

File: cpp/lib/src/util.cc (hunk 13 above)
Comment (mtrojan-ub) [reply cpp/lib/src/util.cc:55]: and what happens with fd_? maybe just call redirectOutput(new_fd) for consistency?

File: cpp/lib/src/util.cc [hunk 16]
```diff
@@ -44,11 +66,14 @@ char *progname; // Must be set in main() with "progname = argv[0];";
 const std::string Logger::FUNCTION_NAME_SEPARATOR(" --> ");
 
 
-
 Logger::Logger()
     : fd_(STDERR_FILENO), log_process_pids_(false), log_no_decorations_(false), log_strip_call_site_(false),
       min_log_level_(LL_INFO)
 {
+    log_fd = fd_;
```
Comment (mtrojan-ub) [inline cpp/lib/src/util.cc:73]: How about calling redirectOutput (fd_) instead?

File: cpp/lib/src/util.cc (hunk 14 above)
Comment (ruschein) [reply cpp/lib/src/util.cc:92]: I did indeed! :-(

File: cpp/lib/src/util.cc (hunk 15 above)
Comment (ruschein) [reply cpp/lib/src/util.cc:42]: Good point.  I eliminated fd_.

File: cpp/lib/src/ProcessUtil.cc (hunk 7 above)
Comment (ruschein) [reply cpp/lib/src/ProcessUtil.cc:35]: That documentation is incorrect! :-( (I fixed it.)

File: cpp/log_rotate.cc (hunk 10 above)
Comment (ruschein) [reply cpp/log_rotate.cc:109]: FileUtil::OnlyKeepLastNLines calls RemoveLeadingBytes which opens the file with O_RDWR before calling ftruncate.
As far as the locking is concerned that is my reading of what the underlying fcntl documentation states.

File: cpp/lib/src/FileLocker.cc (hunk 11 above)
Comment (ruschein) [reply cpp/lib/src/FileLocker.cc:53]: That's the alarm call.

File: cpp/lib/include/FileUtil.h [hunk 17]
```diff
@@ -127,7 +127,7 @@ class Directory {
     public:
         Entry(const Entry &other);
 
-        /** \return the full name of the entry, i.e., including the directory path. */
+        /** \return the name of the entry w/o the directory path. */
```
Comment (mtrojan-ub) [inline cpp/lib/include/FileUtil.h:130]: Then it might be good if we had a GetFullName() function, so we do not need to build the full name again by hand.

File: git-config/hooks/post-merge [hunk 18]
```diff
@@ -46,3 +46,9 @@ if [[ $compare_against_installer_schema && -r /tmp/new_ub_tools.schema ]]; then
         echo "Verification successful! (The current DB schema was successfully compared to the installer's DB schema.)"
     fi
 fi
+
+
+SYSTEM_UPDATER=/usr/local/bin/system_updater
+if [ -e "$SYSTEM_UPDATER" ]; then
+    "$SYSTEM_UPDATER"
```
Comment (mtrojan-ub) [inline git-config/hooks/post-merge:53]: MYSQL_PATCHER takes the directory $MYSQL_SCHEMA_UPDATES_PATH as argument.
How about implementing SYSTEM_UPDATER in the same way?

File: cpp/log_rotate.cc (hunk 9 above)
Comment (ruschein) [reply cpp/log_rotate.cc]: Fixed.

File: cpp/lib/src/util.cc (hunk 16 above)
Comment (ruschein) [reply cpp/lib/src/util.cc]: Obsolete because fd_ does not exist any longer.

File: cpp/lib/include/FileUtil.h (hunk 17 above)
Comment (ruschein) [reply cpp/lib/include/FileUtil.h:130]: It exist now.

File: git-config/hooks/post-merge (hunk 18 above)
Comment (ruschein) [reply git-config/hooks/post-merge]: Done!

File: cpp/lib/src/ProcessUtil.cc [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+std::unordered_set<pid_t> GetProcessIdsForPath(const std::string &path) {
+    std::unordered_set<pid_t> pids;
+
+    FileUtil::Directory proc("/proc", "\\d+"); // Only PID's
+    for (const auto &pid_dir : proc) {
+        const std::string FD_DIR_PATH(pid_dir.getFullName() + "/fd");
+        FileUtil::Directory fd_dir(FD_DIR_PATH);
+        for (const auto &link : fd_dir) {
+            std::string link_target;
+            if (FileUtil::ReadLink(FD_DIR_PATH + "/" + link.getName(), &link_target) and link_target == path) {
```
Comment (mtrojan-ub) [inline cpp/lib/src/ProcessUtil.cc:39]: getFullName() could be used here as well

File: cpp/system_updater.cc [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    if (script_names.empty()) {
+        LOG_INFO("nothing to be done!");
+        return EXIT_SUCCESS;
+    }
+
+    std::sort(script_names.begin(), script_names.end(), ScriptLessThan);
+
+    for (const auto &script_name : script_names)
+        ExecUtil::ExecOrDie(SYSTEM_UPDATES_DIR + "/" + script_name);
```
Comment (mtrojan-ub) [inline cpp/system_updater.cc:89]: Maybe some more LOG_INFO calls might be useful
- when executing a script
- when completing the operation / writing highest_version_number to VERSION_PATH

File: cpp/lib/include/SysLog.h [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    // \brief Write a log entry.
+    void log(const LogLevel level, const std::string &message);
+
+    // \return An or'rd together set of the current log levels.
+    int getLogLevels() const { return ::setlogmask(0); }
+
+    // \brief  Sets a new set of levels that will be logged unless "new_levels" is 0.
+    // \param  new_levels An or'rd together set of the levels that wile be logged after the call.
+    // \return An or'rd together set of the current log levels..
+    int setLogLevsls(const int new_levels) { return ::setlogmask(new_levels); }
```
Comment (mtrojan-ub) [inline cpp/lib/include/SysLog.h:56]: typo => setLogLevsls

File: cpp/data/rsyslog.conf [hunk 22]
```diff
@@ -0,0 +1 @@
+local0.*  /usr/local/var/log/tuefind/syslog.log
```
Comment (mtrojan-ub) [inline cpp/data/rsyslog.conf:1]: please move conf to cpp/data/installer + rename e.g. to syslog.ub_tools.conf

File: cpp/data/system_updates/1.sh [hunk 23]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+install --owner=root --group=root --mode=644 /usr/local/ub_tools/cpp/data/rsyslog.conf /etc/rsyslog.d/ub_tools.conf
```
Comment (mtrojan-ub) [inline cpp/data/system_updates/1.sh:2]: this should also be performed by installer.cc & system_version should be detected by installer

File: cpp/lib/src/SysLog.cc [hunk 24]
```diff
@@ -0,0 +1,38 @@
+/** \file    SysLog.h
```
Comment (mtrojan-ub) [inline cpp/lib/src/SysLog.cc:1]: This is SysLog.cc, not SysLog.h

What's the meaning of the \file tag except redundant information? Guess i will never understand.

File: cpp/lib/src/SysLog.cc [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+ *  \brief   Declaration of class SysLog.
```
Comment (mtrojan-ub) [inline cpp/lib/src/SysLog.cc:2]: I guess this should be "Implementation" and not "Declaration", which is also redundant because we know it due to the file extension...

File: cpp/syslog.cc [hunk 26]
```diff
@@ -0,0 +1,123 @@
+/** \brief A tool for logging to syslogd in a shell script.
```
Comment (mtrojan-ub) [inline cpp/syslog.cc:1]: Could there be conflicts with any other program named syslog in PATH?

File: cpp/log_rotate.cc (hunk 10 above)
Comment (mtrojan-ub) [reply cpp/log_rotate.cc:109]: Again: Why is the flag called WRITE_ONLY if, in fact, we use it to both reading + writing?

File: cpp/log_rotate.cc [hunk 27]
```diff
@@ -117,6 +127,12 @@ int Main(int argc, char *argv[]) {
                                                FileUtil::GroupnameFromGID(gid));
                     if (SELinuxUtil::IsEnabled())
                         SELinuxUtil::FileContext::ApplyChanges(filename);
+
+                    const pid_t our_pid(::getpid());
+                    for (const auto pid : pids) {
+                        if (pid != our_pid)
```
Comment (mtrojan-ub) [inline cpp/log_rotate.cc:133]: GetProcessIdsForPath() sould take an additional bool to exclude the current pid, then we wouldn't have to test for our_pid here explicitly

File: cpp/syslog.cc (hunk 26 above)
Comment (ruschein) [reply cpp/syslog.cc:1]: No, if at all the demon is called syslogd and, I checked, there is no system tool called syslog.

File: cpp/merge_print_and_online.cc [hunk 28]
```diff
@@ -711,7 +711,9 @@ bool MergeFieldPair245(const MARC::Record::Field &merge_field, const MARC::Recor
         if (FuzzyEqual(varying_title_field.getFirstSubfieldWithCode('a'), import_title))
             return true;
     }
-    merge_record->insertField("246", { { 'a', import_title },
+
+    const bool import_title_is_newer(import_record.getPublicationYear() > merge_record->getPublicationYear());
+    merge_record->insertField("246", { { 'a', import_title_is_newer? import_title : merge_title },
```
Comment (mtrojan-ub) [inline cpp/merge_print_and_online.cc:716]: Missing space before '?'
